src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -1956,6 +1968,568 @@ public Task OpenAsync(CancellationToken token)
+                        {
+                            if (!(_isInputClosed))
+                            {
+                                Fx.Assert("Shutdown request was not received.");
+                                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.ShutdownRequestWasNotReceived)));
+                            }
+                        }
+                        return (true, wasAborted);
+                    }
+                }
+                catch (CommunicationObjectAbortedException)
+                {
+                    if (State != CommunicationState.Closed)
+                    {
+                        throw;
+                    }
+                    wasAborted = true;
+                }
+                return (false, wasAborted);
+            }
+
+            private class SoapSecurityServerDuplexSession : SoapSecurityInputSession, IDuplexSession
+            {
+                private ServerSecurityDuplexSessionChannel _channel;
+
+                public SoapSecurityServerDuplexSession(SecurityContextSecurityToken sessionToken, SecuritySessionServerSettings settings, ServerSecurityDuplexSessionChannel channel)
+                    : base(sessionToken, settings, channel)
+                {
+                    _channel = channel;
+                }
+
+                public void CloseOutputSession()
+                {
+                    CloseOutputSession(ServiceDefaults.CloseTimeout);
+                }
+
+                public void CloseOutputSession(TimeSpan timeout)
+                {
+                   // channel.ThrowIfFaulted();
+                   // channel.ThrowIfNotOpened();
+                    Exception pendingException = null;
+                    try
+                    {
+                        _channel.CloseOutputSessionAsync(new TimeoutHelper(timeout).GetCancellationToken());
+                    }
+                    catch (Exception e)
+                    {
+                        if (Fx.IsFatal(e))
+                        {
+                            throw;
+                        }
+                        pendingException = e;
+                    }
+                    if (pendingException != null)
+                    {
+                        _channel.OnFaulted(pendingException);
+                        if (pendingException is CommunicationException)
+                        {
+                            throw pendingException;
+                        }
+                        else
+                        {
+                            throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(pendingException);
+                        }
+                    }
+                }
+
+                public Task CloseOutputSessionAsync()
+                {
+                    CloseOutputSession(ServiceDefaults.CloseTimeout); ;
+                    return Task.CompletedTask;
```

Comment:
This method and the following one shouldn't exist as the two methods above will becoming async and will replace them.

Reply:
I have removed one unnecessary method and made one private (to avoid confusion).. but these two methods needs to be there to respect IDuplexSession interface.